# Ray Fair — Econometrics and Presidential Elections: Two applicaitons

The following jupyter notebook attempts to better understand Prof. Fair's work on forecasting elections (Fair, 1996, 2007) by appying his models and ideas to other elections. Fair's model is unique in the sense that it uses relitavely simple predictors (no fancy machine learning), which are macroeconomic in nature (no polling data, etc.). His model is known for its predictive power and therefore has served as an inspiration for this project. I am interested to what extend I can use his ideas in other contexts and what modifications different entities require. His model is very much tailored to the US electorial system and as you will see has limitations in other contexts. The model works quite well for the US due to it's bypartisan system and the presidents (assumed) power to affect macroeconomic indicators. 

The following model is the one specified in Prof. Ray's work:

**The model**

$$V^{p} = \alpha_{0} + \alpha_{1} G + \alpha_{2} P + \alpha_{3} Z + \alpha_{4} I + \alpha_{5} \mathrm{DUR} + \alpha_{6} \mathrm{DPER} + \alpha_{7} \mathrm{WAR} + \varepsilon$$

**Variables**

| Variable | Definition |
|---|---|
| `V^p` | Democratic share of the two-party presidential vote. |
| `G` | Growth rate of real per-capita GDP in the first 3 quarters of the on-term election year (annual rate). |
| `P` | Absolute value of the growth rate of the GDP deflator in the first 15 quarters of the administration (annual rate), except 1920, 1944, 1948 where values are zero. |
| `Z` | Number of quarters in the first 15 quarters where real per-capita GDP growth is &gt; 3.2% (annual rate), except 1920, 1944, 1948 where zero. |
| `I` | 1 if Democratic presidential incumbent at election; −1 if Republican. |
| `DUR` | 0 if either party has been in the White House for one term; 1 [−1] for two consecutive Democratic [Republican] terms; 1.25 [−1.25] for three terms; 1.50 [−1.50] for four terms; etc. |
| `DPER` | 1 if a Democratic presidential incumbent runs again; −1 if a Republican incumbent runs again; 0 otherwise. |
| `WAR` | 1 for the elections of 1918, 1920, 1942, 1944, 1946, 1948; 0 otherwise. |






I chose the first application out of relevance. The New York City Mayor Elections were coming up and I wanted to use Prof. Fair's model to explain election outcomes. It hoped that it would have help me to understand whether (national) economic indicators influence local voter behaviour. The main problem is the limitation of (historic) data. It became obvious that detailed data sets on local economic data for NYC are hard to find. 
I am aware of the limitations of my work and with more resources and further work, I could maybe have found better data or could have potentially predicted it. The imperfections of this project are obvious to me but that's not the point, I wanted to rather use this as good practice and a way to demonstrate my research skills while also better understanding Prof. Ray's work, which I find increasingly intriguing, also as I embark on my journey to find a specific research interest for my future academic career (seminars, senior thesis, Phd, etc.). In that sense, this project might be a starting point for either thinking about economic models for predicting local elections, a project on causally explaining voting behaviour (to what extend are voters influenced by federal policy in local elections) or the application of Prof. Fair's model to other countries, like my own, Germany. Next, I will apply the model with step by step applicaitons. 



First, I import data on NYC mayoral election outcomes, sourced from wikipedia. This data gives me a table with all the mayoral candidates, the voting year and their results. Inspired by Prof. Fair, I'll use the percentage for the democratic candidate as my dependent variable. One issue I encountered and should put more thought into is the problem of how independent candidates that are in between conservative and progressive might affect the model. In 1913 and 2009 an independent candidate won the election, which the model should account for. Otherwise, the New York City mayoral elections have a similar bypartisan structure or divide, which Prof. Fair's model also builds on. 




In [505]:
import pandas as pd

NYC = pd.read_csv("nyc_mayoral_elections_1897_2025.csv")
NYC.head(100)


year  total_votes                               democratic_candidate  \
0   1897     523560.0                                 Robert A. Van Wyck   
1   1901     579301.0                                  Edward M. Shepard   
2   1903     589898.0                            George B. McClellan Jr.   
3   1905     604673.0                            George B. McClellan Jr.   
4   1909     594902.0                                  William J. Gaynor   
5   1913     627017.0                                      Edward McCall   
6   1917     673300.0                                 John Francis Hylan   
7   1921    1168767.0                          John Francis Hylan (inc.)   
8   1925    1137966.0                                       Jimmy Walker   
9   1929    1429385.0                                Jimmy Walker (inc.)   
10  1932    1990899.0                                    John P. O'Brien   
11  1933    2152505.0                             John P. O'Brien (inc.)   
12  1937    2235386.0  Jeremiah T. Mahoney, D-Trades Union-Anticommunist   
13  1941    2263369.0                                    William O'Dwyer   
14  1945    1982361.0                             William O'Dwyer, D-ALP   
15  1949    2591684.0                             William O'Dwyer (inc.)   
16  1950    2626476.0                            Ferdinand Pecora, D-Lib   
17  1953    2204368.0                                      Robert Wagner   
18  1957    2179878.0                 Robert Wagner, D-Lib-Fusion (inc.)   
19  1961    2424985.0            Robert Wagner, D-Lib-Brotherhood (inc.)   
20  1965    2554199.0              Abraham Beame, D-Civil Service Fusion   
21  1969    2390834.0          Mario Procaccino, D-Ind-Civil Service Ind   
22  1973    1705634.0                                      Abraham Beame   
23  1977    1435113.0                                            Ed Koch   
24  1981    1222644.0                                Ed Koch, D-R (inc.)   
25  1985    1112796.0                              Ed Koch, D-Ind (inc.)   
26  1989    1819260.0                                      David Dinkins   
27  1993    1783937.0                               David Dinkins (inc.)   
28  1997    1116358.0                                     Ruth Messinger   
29  2001    1480582.0                                  Mark Green, D-WFP   
30  2005    1289935.0                                    Fernando Ferrer   
31  2009    1154802.0                               Bill Thompson, D-WFP   
32  2013    1087710.0                              Bill de Blasio, D-WFP   
33  2017    1166314.0                       Bill de Blasio, D-WFP (inc.)   
34  2021    1125258.0                                         Eric Adams   
35  2025          NaN                              Zohran Mamdani, D-WFP   

    democratic_votes  democratic_pct  \
0           233997.0            44.7   
1           265177.0            45.8   
2           314782.0            53.4   
3           228407.0            37.8   
4           250378.0            42.1   
5           233919.0            37.3   
6           314010.0            46.6   
7           750247.0            64.2   
8           748687.0            65.8   
9           867522.0            60.7   
10         1054324.0            53.0   
11          586672.0            27.3   
12          890756.0            39.8   
13         1054235.0            46.6   
14         1125357.0            55.3   
15         1266512.0            48.9   
16          935351.0            35.6   
17         1022626.0            46.3   
18         1509775.0            69.2   
19         1237421.0            51.0   
20         1046699.0            41.0   
21          831772.0            34.8   
22          963542.0            56.5   
23          717376.0            50.0   
24          912622.0            74.6   
25          868260.0            78.0   
26          917544.0            50.4   
27          858868.0            48.1   
28          479288.0           

When I started this project, it was a reasonable assumption to me that federal economic data would not be a good predictor for local election data, since a New York City mayor usually has little power to influence macroeconomic policy, unless you are afraid of his power to influence wall street by closing down the metro or anyother reasonable hypothesis. Hence, I went on the search for local economic data that I believed to be much more relevant for local voter behaviour like local unemployment rates, income, etc.. Rent prices for example are usually subject of discussion in local election campaigns and affect many New Yorks. Additionally, mayors can to an extend actually through rent controls affect this economic indicator. Unfortunately, my search for local economic data was unsuccessful. While I did find much data on the FED of St. Luis website, I had to accept the limitation of historic quarterly data. This made it unfortunately impossible for me to use lcoal economic data, since I am reliant on historic data due to the small number of data points due to the 4 year gap. The FED of St. Louis actually offers local CPI and the Rent price index for new york, which dates back to 1914. However, because of the NA's the CPI data is only usable after 1945 and the rent data even later. Hence, I had to come to the conclusion that I might not be able to use local economic data effectively. Moreover, I didn't have any local income measure for G.

However, the usage of federal income data provided an equally interesting scope by responding to the question of how federal economic data actually affects local voting behaviour or more generally how federal politics affects local voting behaviour. 



Here, I download local CPI and Rent data for NYC, which I will not be able to use unfortunately.

In [506]:
import pandas as pd

CPI = pd.read_csv("CUURA101SA0.csv")
CPI.head(100)

Rent = pd.read_csv("CUURA101SEHA.csv")
Rent.head(100)

observation_date  CUURA101SEHA
0        1914-12-01          18.9
1        1915-01-01           NaN
2        1915-02-01           NaN
3        1915-03-01           NaN
4        1915-04-01           NaN
..              ...           ...
95       1922-11-01           NaN
96       1922-12-01          29.7
97       1923-01-01           NaN
98       1923-02-01           NaN
99       1923-03-01          30.0

[100 rows x 2 columns]

Since I couldn't find good historic local income economic data for NYC, I had to utilize national income data for G. Hence, I can look at the role of national economic indicators in local elections.

Here, I import the quarterly GDP data from Prof. Fair's website for the whole US.

In [507]:
import pandas as pd

# Example colspecs: adjust (start, end) to your file’s layout
colspecs = [(0, 7), (9, 18), (19, 28), (30, 38)]
GDP_quarterly = pd.read_fwf(
    "Quarterly GDP Data.txt",
    colspecs=colspecs,
    names=["Quarter", "(Y)", "(X)", "(P)"],
    header=None
)
GDP_quarterly = GDP_quarterly.iloc[3:]
GDP_quarterly.tail(100)


Quarter       (Y)       (X)      (P)
484  1997.4   8765.90  11722.70  274.246
485  1998.1   8866.50  11839.90  274.950
486  1998.2   8969.70  11949.50  275.703
487  1998.3   9121.10  12099.20  276.564
488  1998.4   9294.00  12294.70  277.400
..      ...       ...       ...      ...
579  2021.3  23550.40  19672.60  332.297
580  2021.4  24349.10  20006.20  332.584
581  2022.1  24740.50  19924.10  332.749
582  2022.2  25248.50  19895.30  332.940
583  2022.3  25663.30  20021.70  333.241

[100 rows x 4 columns]

NYC Mayoral Elections are always held in November, so I will always take the first 3 quarters of the election year to calculate G. In Prof. Fair's model, G is the growth rate of Real GDP per capita in the first 3 quarters before the election. For the NYC election that's Q1-Q3 of the specific election year. The following code, creates variable G: 



In [508]:
# Create Real GDP per Capita variable
GDP_quarterly['Real GDP per Capita'] = (GDP_quarterly['(X)'].astype(float)*100)/ GDP_quarterly['(P)'].astype(float)

# Get rid of years before 1914
GDP_quarterly = GDP_quarterly[GDP_quarterly['Quarter'].str.slice(0,4).astype(int) >= 1914]

# Get rid of all years that are not mayoral election years: 1917 1921 1925 1929 1932 1933 1937 1941 1945 1949 1950 1953 1957 1961 1965 1969 1973 1977 1981 1985 1989 1993 1997 2001 2005 2009 2013 2017 2021
GDP_quarterly = GDP_quarterly[GDP_quarterly['Quarter'].str.slice(0,4).astype(int).isin([1917, 1921, 1925, 1929, 1932, 1933, 1937, 1941, 1945, 1949, 1950, 1953, 1957, 1961, 1965, 1969, 1973, 1977, 1981, 1985, 1989, 1993, 1997, 2001, 2005, 2009, 2013, 2017, 2021])]

# Calculate annual percentage change in Real GDP per Capita for first three quarters of each mayoral election year, i.e. taking the percentage change from Q1 to Q3 of the election year and extrapolating to an annual rate by taking it to the power of 4/3

# Extract year + quarter number
GDP_quarterly["Year"] = GDP_quarterly["Quarter"].astype(str).str.split(".").str[0].astype(int)
GDP_quarterly["Q"] = GDP_quarterly["Quarter"].astype(str).str.split(".").str[1].astype(int)

# Pivot to wide format: each year gets Q1, Q3 columns
wide = GDP_quarterly.pivot(index="Year", columns="Q", values="Real GDP per Capita")

# Ensure needed quarters exist
wide = wide.rename(columns={1: "Q1", 3: "Q3"})

# Compute G
wide["G"] = ((wide["Q3"] / wide["Q1"])**(4/3) - 1) * 100

# Merge back if needed
GDP_quarterly = GDP_quarterly.merge(
    wide["G"],
    on="Year",
    how="left"
)
GDP_quarterly.head(100)


G = GDP_quarterly[~GDP_quarterly["Q"].isin([1,2,3])].reset_index(drop=True)

GDP_quarterly.head(50)

G = G.iloc[:, 1:]
G.head(15)

(Y)      (X)      (P)  Real GDP per Capita  Year  Q          G
0    64.89   772.07  103.445           746.357968  1917  4   5.128574
1    70.37   734.54  109.267           672.243221  1921  4   6.290097
2    96.44  1000.49  116.522           858.627555  1925  4   4.125995
3   102.07  1083.95  122.377           885.746505  1929  4   5.671940
4    55.67   798.63  125.231           637.725483  1932  4 -11.149834
5    59.53   817.01  125.974           648.554464  1933  4  21.030681
6    87.28  1111.06  129.304           859.261894  1937  4   2.432873
7   143.23  1667.40  133.904          1245.220456  1941  4  14.373973
8   213.70  2116.91  140.497          1506.729681  1945  4 -11.378254
9   270.60  2103.70  150.167          1400.906990  1949  4  -0.187928
10  319.90  2384.90  152.658          1562.250259  1950  4   8.342581
11  386.00  2664.30  160.639          1658.563612  1953  4  -0.821448
12  474.90  2985.80  172.417          1731.731790  1957  4  -0.136632
13  580.60  3440.90  184.894          1861.012256  1961  4   3.779882
14  771.90  4304.70  195.279          2204.384496  1965  4   3.868982

Next, I attempt to create P. I'll use the local CPI data which is standardized at the year 1984. The historic CPI data that is available to me is monthly data which I aggregate to the quarter level to then calculate the average annual inflation for the first 15 quarters in the legislation. Even though local policy might be far from causally related with local inflation, it is interesting to look at it's influence on local voter behaviour. I do the same for the rent price index by creating a variable R. 

In [509]:
import pandas as pd
import numpy as np
import statsmodels.api as sm  # if you run regressions later

# ---------- Helpers ----------
def quarter_start_series(df, date_col, value_col):
    """Return a Series of values at quarter starts (Jan 1, Apr 1, Jul 1, Oct 1)."""
    s = df.copy()
    s[date_col] = pd.to_datetime(s[date_col])
    s = s.sort_values(date_col).set_index(date_col)
    # keep only 01-01, 04-01, 07-01, 10-01
    s = s.loc[s.index.is_month_start & s.index.month.isin([1, 4, 7, 10]), value_col]
    return s.sort_index()

def fair_P_from_level(qstart_level, election_years, anchor="10-01"):
    """
    Compute Fair's P using level series at quarter starts.
    Anchor = start of election quarter (Q4 -> '10-01').
    Requires 16 quarters before the anchor (i.e., idx >= 16).
    """
    qc = qstart_level.to_frame("LEVEL").copy()
    qc["year"] = qc.index.year

    out = []
    for yr in sorted(pd.Series(election_years).astype(int).unique()):
        election_q_start = f"{yr}-{anchor}"
        try:
            idx_e = qc.index.get_loc(pd.Timestamp(election_q_start))
        except KeyError:
            # if exact Oct 1 missing, skip
            continue

        if idx_e < 16:
            continue  # not enough history

        # 1st and 15th quarters of the administration
        LEVEL_1  = qc.iloc[idx_e - 15]["LEVEL"]
        LEVEL_15 = qc.iloc[idx_e - 1]["LEVEL"]

        P = (((LEVEL_15 / LEVEL_1) ** (4 / 15)) - 1) * 100
        out.append({"year": int(yr), "P": float(abs(P))})

    return pd.DataFrame(out)

# ---------- CPI → P ----------
# CPI: columns ['observation_date','CUURA101SA0']
CPI_qstart = quarter_start_series(CPI, "observation_date", "CUURA101SA0")
P_df = fair_P_from_level(CPI_qstart, NYC["year"])

# ---------- CPI Rent → R (same construction as P, separate column) ----------
# Rent: columns ['observation_date','CUURA101SEHA']
Rent_qstart = quarter_start_series(Rent, "observation_date", "CUURA101SEHA")

R_rows = []
qc_r = Rent_qstart.to_frame("LEVEL").copy()

for yr in sorted(NYC["year"].astype(int).unique()):
    election_q_start = pd.Timestamp(f"{yr}-10-01")
    try:
        idx_e = qc_r.index.get_loc(election_q_start)
    except KeyError:
        continue
    if idx_e < 16:
        continue

    L1  = qc_r.iloc[idx_e - 15]["LEVEL"]
    L15 = qc_r.iloc[idx_e - 1]["LEVEL"]
    R = (((L15 / L1) ** (4 / 15)) - 1) * 100
    R_rows.append({"year": int(yr), "R": float(abs(R))})

R_df = pd.DataFrame(R_rows)

# ---------- G was already computed into a year-level table `G` ----------
# Make sure G has columns ['year','G']
if "year" not in G.columns:
    G = G.rename(columns={"Year": "year"})
G_use = G[["year", "G"]].drop_duplicates()

# ---------- Merge everything ONTO NYC (left) ----------
NYC = (NYC
       .merge(G_use, on="year", how="left")
       .merge(P_df,  on="year", how="left")
       .merge(R_df,  on="year", how="left"))

# Optional: sanity checks
# print(NYC[["year","G","P","R"]].sort_values("year"))


Next, I manually create a variable I, for whether the candidate is an incumbent or not, DUR for it's duration in office, and DPER for the duration of the the candidates party in office. 


In [510]:
# Define year → I mapping
I_map = {
    1905: 1,
    1921: 1,
    1929: 1,
    1937: 1,
    1941: 1,
    1949: 1,
    1957: 1,
    1961: 1,
    1981: 1,
    1985: 1,
    1993: 1,
    2017: 1,

    1969: -1,
    1997: -1,
    2005: -1,
    2009: -1
}


NYC["I"] = NYC["year"].map(I_map).fillna(0).astype(int)

# Create DUR variable

DUR_map = {
    1949: 1,
    1957: 1,
    1961: 1.25,

    1977: 1,
    1981: 1.25,
    1985: 1.50,
    1989: 1.75,

    1997: -1,
    2001: -1.25,
    2004: -1.50,
    2009: -1.75,

    2017: 1,
    2021: 1.25
}

NYC["DUR"] = NYC["year"].map(DUR_map).fillna(0)

# Create DPER variable

DPER_map = {



    1949: 1,
    1957: 1,
    1961: 1,

    1969: -1,

    1981: 1,
    1985: 1,

    1997: -1,

    2005: -1,
    2009: -1,

    2017: 1
}

NYC["DPER"] = NYC["year"].map(DPER_map).fillna(0)

NYC.columns

Index(['year', 'total_votes', 'democratic_candidate', 'democratic_votes',
       'democratic_pct', 'major_third_party_candidate', 'third_party_votes',
       'third_party_pct', 'republican_candidate', 'republican_votes',
       'republican_pct', 'other_major_candidates', 'other_votes', 'other_pct',
       'G', 'P', 'R', 'I', 'DUR', 'DPER'],
      dtype='object')

Next, I am fitting the model. The first thing that concerns me is the low number of observations, which is only 21. That's why I have to be careful to not overfit by using too many predictors. I have a degree of freedom 18 for the residuals, which is enough to run OLS. However, I will respect the maximum of 3 predictors to not risk any overfitting. Moreover, to account for autocorrelation in my error-term I use HAC standard errors. In the following model, G_I and R_I plus a constant are my explanatory variables. Both coefficients are positive and statistically significant. However, I am concerned about omitted variable bias and quite a small adjusted R^2 of only 38.9 % of the variance explained, while also having had to omit three variables from Prof. Fair's model in order to not overfit. 

In [511]:
NYC.head(40)

year  total_votes                               democratic_candidate  \
0   1897     523560.0                                 Robert A. Van Wyck   
1   1901     579301.0                                  Edward M. Shepard   
2   1903     589898.0                            George B. McClellan Jr.   
3   1905     604673.0                            George B. McClellan Jr.   
4   1909     594902.0                                  William J. Gaynor   
5   1913     627017.0                                      Edward McCall   
6   1917     673300.0                                 John Francis Hylan   
7   1921    1168767.0                          John Francis Hylan (inc.)   
8   1925    1137966.0                                       Jimmy Walker   
9   1929    1429385.0                                Jimmy Walker (inc.)   
10  1932    1990899.0                                    John P. O'Brien   
11  1933    2152505.0                             John P. O'Brien (inc.)   
12  1937    2235386.0  Jeremiah T. Mahoney, D-Trades Union-Anticommunist   
13  1941    2263369.0                                    William O'Dwyer   
14  1945    1982361.0                             William O'Dwyer, D-ALP   
15  1949    2591684.0                             William O'Dwyer (inc.)   
16  1950    2626476.0                            Ferdinand Pecora, D-Lib   
17  1953    2204368.0                                      Robert Wagner   
18  1957    2179878.0                 Robert Wagner, D-Lib-Fusion (inc.)   
19  1961    2424985.0            Robert Wagner, D-Lib-Brotherhood (inc.)   
20  1965    2554199.0              Abraham Beame, D-Civil Service Fusion   
21  1969    2390834.0          Mario Procaccino, D-Ind-Civil Service Ind   
22  1973    1705634.0                                      Abraham Beame   
23  1977    1435113.0                                            Ed Koch   
24  1981    1222644.0                                Ed Koch, D-R (inc.)   
25  1985    1112796.0                              Ed Koch, D-Ind (inc.)   
26  1989    1819260.0                                      David Dinkins   
27  1993    1783937.0                               David Dinkins (inc.)   
28  1997    1116358.0                                     Ruth Messinger   
29  2001    1480582.0                                  Mark Green, D-WFP   
30  2005    1289935.0                                    Fernando Ferrer   
31  2009    1154802.0                               Bill Thompson, D-WFP   
32  2013    1087710.0                              Bill de Blasio, D-WFP   
33  2017    1166314.0                       Bill de Blasio, D-WFP (inc.)   
34  2021    1125258.0                                         Eric Adams   
35  2025          NaN                              Zohran Mamdani, D-WFP   

    democratic_votes  democratic_pct  \
0           233997.0            44.7   
1           265177.0            45.8   
2           314782.0            53.4   
3           228407.0            37.8   
4           250378.0            42.1   
5           233919.0            37.3   
6           314010.0            46.6   
7           750247.0            64.2   
8           748687.0            65.8   
9           867522.0            60.7   
10         1054324.0            53.0   
11          586672.0            27.3   
12          890756.0            39.8   
13         1054235.0            46.6   
14         1125357.0            55.3   
15         1266512.0            48.9   
16          935351.0            35.6   
17         1022626.0            46.3   
18         1509775.0            69.2   
19         1237421.0            51.0   
20         1046699.0            41.0   
21          831772.0            34.8   
22          963542.0            56.5   
23          717376.0            50.0   
24          912622.0            74.6   
25          868260.0            78.0   
26          917544.0            50.4   
27          858868.0            48.1   
28          479288.0           

In [512]:
import pandas as pd
import numpy as np
import statsmodels.api as sm



# 1) Build X, y
X = NYC[['G', 'P']].astype(float)
y = NYC['democratic_pct'].astype(float)

df = pd.concat([X, y], axis=1)
df = df.replace([np.inf, -np.inf], np.nan).dropna()

X_clean = df[['G', 'P']]
y_clean = df['democratic_pct']

# 2) Add constant and fit, drop rows with any NaN in X or y, and infinite values
X_clean = sm.add_constant(X_clean, has_constant='add')
nyc_reg = sm.OLS(y_clean, X_clean).fit(cov_type='HAC',cov_kwds={'maxlags':1})
print(nyc_reg.summary())


                            OLS Regression Results                            
Dep. Variable:         democratic_pct   R-squared:                       0.037
Model:                            OLS   Adj. R-squared:                 -0.070
Method:                 Least Squares   F-statistic:                    0.8468
Date:                Sun, 23 Nov 2025   Prob (F-statistic):              0.445
Time:                        17:44:49   Log-Likelihood:                -82.813
No. Observations:                  21   AIC:                             171.6
Df Residuals:                      18   BIC:                             174.8
Df Model:                           2                                         
Covariance Type:                  HAC                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         52.2635      6.580      7.942      0.0

Here I define the basic model with G and I:

In [513]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

# 1) Build X, y
X = NYC[['G', 'P']].astype(float)
y = NYC['democratic_pct'].astype(float)

# 2) Combine and clean (remove NaN and ±inf)
df = pd.concat([X, y], axis=1)
df = df.replace([np.inf, -np.inf], np.nan).dropna()

# 3) Split into train (all but last) and test (last row only)
train = df.iloc[:-1]
test  = df.iloc[[-1]]   # keep as DataFrame

X_train = train[['G', 'P']]
y_train = train['democratic_pct']

X_test = test[['G', 'P']]
y_test_actual = test['democratic_pct'].iloc[0]   # actual last value

# 4) Add constant
X_train = sm.add_constant(X_train, has_constant='add')
X_test  = sm.add_constant(X_test, has_constant='add')

# 5) Fit model on training set only
nyc_reg = sm.OLS(y_train, X_train).fit(cov_type='HAC', cov_kwds={'maxlags': 1})
print(nyc_reg.summary())

# 6) Predict the last observation
y_pred_last = nyc_reg.predict(X_test)

print("\nActual last democratic_pct:", y_test_actual)
print("Predicted last democratic_pct:", float(y_pred_last))


                            OLS Regression Results                            
Dep. Variable:         democratic_pct   R-squared:                       0.059
Model:                            OLS   Adj. R-squared:                 -0.052
Method:                 Least Squares   F-statistic:                     1.375
Date:                Sun, 23 Nov 2025   Prob (F-statistic):              0.280
Time:                        17:44:49   Log-Likelihood:                -78.540
No. Observations:                  20   AIC:                             163.1
Df Residuals:                      17   BIC:                             166.1
Df Model:                           2                                         
Covariance Type:                  HAC                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         50.7739      6.645      7.641      0.0

/var/folders/x5/gq2pkkwj427ctp4rg1stnj180000gn/T/ipykernel_70400/4102661690.py:35: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  print("Predicted last democratic_pct:", float(y_pred_last))


Here I define a more extensive model with G_I and P_I and all its interaction terms. 

In [514]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

# 1) Build X, y (need G and I to form G×I)
X = NYC[['G', 'P', 'R', 'I', 'DUR']].astype(float)
y = NYC['democratic_pct'].astype(float)

# 2) Create interaction term G × I
X['G_I'] = X['G'] * X['I']
X['P_I'] = X['P'] * X['I']
X['R_I'] = X['R'] * X['I']

# 3) Keep only the interaction term in the regression
X = X[['G', 'G_I', 'P', 'P_I', 'R_I', 'I', 'DUR']]

# 4) Combine X and y, remove NA and ±inf rows
df = pd.concat([X, y], axis=1)
df = df.replace([np.inf, -np.inf], np.nan).dropna()

# 5) Re-split cleaned data
X_clean = df[['G', 'P', 'G_I', 'P_I', 'I']]    
y_clean = df['democratic_pct']

# 6) Add constant and fit OLS with HAC (Newey-West)
X_clean = sm.add_constant(X_clean, has_constant='add')
nyc_reg = sm.OLS(y_clean, X_clean).fit(cov_type='HAC', cov_kwds={'maxlags': 1})

print(nyc_reg.summary()) 


                            OLS Regression Results                            
Dep. Variable:         democratic_pct   R-squared:                       0.579
Model:                            OLS   Adj. R-squared:                  0.278
Method:                 Least Squares   F-statistic:                     9.216
Date:                Sun, 23 Nov 2025   Prob (F-statistic):            0.00550
Time:                        17:44:49   Log-Likelihood:                -47.171
No. Observations:                  13   AIC:                             106.3
Df Residuals:                       7   BIC:                             109.7
Df Model:                           5                                         
Covariance Type:                  HAC                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         74.0213     12.068      6.134      0.0

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=13
  res = hypotest_fun_out(*samples, **kwds)


Here I predict the election result for 2021 using a model trained on data not inlcuding 2021 and compare it to the actual value: 

In [515]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

# 1) Build X, y (need G, P, R, I, DUR to form interactions)
X = NYC[['G', 'P', 'R', 'I', 'DUR']].astype(float)
y = NYC['democratic_pct'].astype(float)

# 2) Create interaction terms
X['G_I'] = X['G'] * X['I']
X['P_I'] = X['P'] * X['I']
X['R_I'] = X['R'] * X['I']

# 3) Keep intended predictors
X = X[['G', 'P', 'G_I', 'P_I', 'I', 'DUR', 'R_I']]

# 4) Combine X and y, remove NA and ±inf rows
df = pd.concat([X, y], axis=1)
df = df.replace([np.inf, -np.inf], np.nan).dropna()

# 5) Train/test split: all but last row for training
train = df.iloc[:-1]
test  = df.iloc[[-1]]          # keep as DataFrame

X_train = train[['G', 'P', 'G_I', 'P_I', 'I']]
y_train = train['democratic_pct']

X_test  = test[['G', 'P', 'G_I', 'P_I', 'I']]
y_test_actual = test['democratic_pct'].iloc[0]

# 6) Add constant
X_train = sm.add_constant(X_train, has_constant='add')
X_test  = sm.add_constant(X_test, has_constant='add')

# 7) Fit model on training data only (HAC standard errors)
nyc_reg = sm.OLS(y_train, X_train).fit(cov_type='HAC', cov_kwds={'maxlags': 1})
print(nyc_reg.summary())

# 8) Predict last observation
y_pred_last = nyc_reg.predict(X_test)

print("\nActual last democratic_pct:", y_test_actual)
print("Predicted last democratic_pct:", float(y_pred_last))


                            OLS Regression Results                            
Dep. Variable:         democratic_pct   R-squared:                       0.617
Model:                            OLS   Adj. R-squared:                  0.298
Method:                 Least Squares   F-statistic:                     8.239
Date:                Sun, 23 Nov 2025   Prob (F-statistic):             0.0116
Time:                        17:44:49   Log-Likelihood:                -43.077
No. Observations:                  12   AIC:                             98.15
Df Residuals:                       6   BIC:                             101.1
Df Model:                           5                                         
Covariance Type:                  HAC                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         69.3172     12.364      5.606      0.0

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)
/var/folders/x5/gq2pkkwj427ctp4rg1stnj180000gn/T/ipykernel_70400/2583133482.py:43: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  print("Predicted last democratic_pct:", float(y_pred_last))


Here I employ Leave One Out Cross Validation to assess the predictive power of my models:

In [540]:
from sklearn.model_selection import LeaveOneOut
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

# 1) Build X, y (need G and I to form G×I)
X = NYC[['G', 'P', 'I']].astype(float).dropna()
y = NYC['democratic_pct'].astype(float).dropna()
# 2) Create interaction term G × I
X['G_I'] = X['G'] * X['I']
X['P_I'] = X['P'] * X['I']

# 3) Keep only the interaction term in the regression
X = X[['G_I', 'P_I']]

loo = LeaveOneOut()
preds = []
truth = []

for train_idx, test_idx in loo.split(X):
    X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
    y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]

    model = LinearRegression(fit_intercept=True)
    model.fit(X_train, y_train) 

    preds.append(model.predict(X_test)[0])
    truth.append(y_test.values[0])

# Compute LOOCV error
preds = np.array(preds)
truth = np.array(truth)

rmse = np.sqrt(mean_squared_error(truth, preds))
print("LOOCV RMSE:", rmse)

# Optional R^2 on LOOCV predictions
r2 = 1 - np.sum((truth - preds)**2) / np.sum((truth - truth.mean())**2)
print("LOOCV R^2:", r2)


LOOCV RMSE: 10.826111103044418
LOOCV R^2: -0.07917691321151521


## German Parliamentary Elections

New York City's case has proven to be difficult but I'd still like to use Prof. Fair's model for an application other than the US presidential election. I decided to look at German federal elections, if I deem it to be reasonable to use a modification of his model and if the data looks good, I'll attempt to predict German elections. 

# 1. 

Prof. Fair with his model predicts the percentage of the democratic candidate in the presedential elections. Since the US system is basically a two-party system, his choice of dependent variable proofs to work well to predict presedential election outcomes in the US. In contrast, Germany is a multi-party system and the chancellor gets elected by the governing coalition. Predicting coalitions is relatively hard because of the qualitative data it depends on but maybe we can use Prof. Fair's model to predict vote shares of specific parties. Another approach could be to ignore smaller parties and only focus on the christian democracts (conservatives) and the social democrats (progressives) as the two main forces in German politics. This was a reasonable assumption up until the 2010s when the German party system became more and more fragmented and the two dominating bipolar forces lost vote shares. To solve this, it could make sense to cluster parties in order to have two blocs and a bipolar system comparable to the US. An easier way that also builds upon the idea of Prof. Fair's model to look at how incumbent candidates or parties obtain credibility for their eocnomic policy, let's predict the governing party's vote share (i.e. the party that nominates the chancellor and the incumbent).

# 2. 

Following my decision to predict the governing party's vote share, I need a data set with the governing party's vote share for each election. 
I take that from the website of the german Bundestag with the first election in 1949 and the last in 2025:




In [517]:
# Importing election data: 

import pandas as pd 

results = pd.read_excel("German Elections.xlsx", header=1)





In [518]:
results.head(100)

Jahr  CDU/CSU   SPD   FDP  Die Grünen  Bündnis 90/Die Grünen  \
0   2025     28.6  16.4   4.3         NaN                   11.6   
1   2021     24.1  25.7  11.5         NaN                   14.8   
2   2017     32.9  20.5  10.7         NaN                    8.9   
3   2013     41.5  25.7   4.8         NaN                    8.4   
4   2009     33.8  23.0  14.6         NaN                   10.7   
5   2005     35.2  34.2   9.8         NaN                    8.1   
6   2002     38.5  38.5   7.4         NaN                    8.6   
7   1998     35.2  40.9   6.2         NaN                    6.7   
8   1994     41.5  36.4   6.9         NaN                    7.3   
9   1990     43.8  33.5  11.0         3.8                    1.2   
10  1987     44.3  37.0   9.1         8.3                    NaN   
11  1983     48.8  38.2   7.0         5.6                    NaN   
12  1980     44.5  42.9  10.6         1.5                    NaN   
13  1976     48.6  42.6   7.9         NaN                    NaN   
14  1972     44.9  45.8   8.4         NaN                    NaN   
15  1969     46.1  42.7   5.8         NaN                    NaN   
16  1965     47.6  39.3   9.5         NaN                    NaN   
17  1961     45.3  36.2  12.8         NaN                    NaN   
18  1957     50.2  31.8   7.7         NaN                    NaN   
19  1953     45.2  28.8   9.5         NaN                    NaN   
20  1949     31.0  29.2  11.9         NaN                    NaN   

    Die Linke. PDS   AfD  SSW  Sonstige  
0              8.8  20.8  0.2       9.4  
1              4.9  10.3  0.1       8.7  
2              9.2  12.6  NaN       5.0  
3              8.6   4.7  NaN       6.3  
4             11.9   NaN  NaN       6.0  
5              8.7   NaN  NaN       4.0  
6              4.0   NaN  NaN       3.0  
7              5.1   NaN  NaN       5.9  
8              4.4   NaN  NaN       3.5  
9              2.4   NaN  NaN       4.3  
10             NaN   NaN  NaN       1.3  
11             NaN   NaN  NaN       0.4  
12             NaN   NaN  NaN       0.5  
13             NaN   NaN  NaN       0.9  
14             NaN   NaN  NaN       0.9  
15             NaN   NaN  NaN       5.5  
16             NaN   NaN  NaN       3.6  
17             NaN   NaN  NaN       5.7  
18             NaN   NaN  NaN      10.5  
19             NaN   NaN  NaN      16.5  
20             NaN   NaN  NaN      27.9

Some data cleaning:

In [519]:
import pandas as pd
import numpy as np

# 1) Drop 1949 and keep only relevant years
df = results[results['Jahr'] >= 1953].copy()

# 2) Choose the party per year (your rules)
c1 = df['Jahr'].between(1953, 1969)     # Union
c2 = df['Jahr'].between(1972, 1983)     # SPD
c3 = df['Jahr'].between(1987, 1998)     # Union
c4 = df['Jahr'].between(2002, 2005)     # SPD
c5 = df['Jahr'].between(2009, 2021)     # Union
c6 = df['Jahr'] == 2025                 # SPD

df['party'] = np.select(
    [c1, c2, c3, c4, c5, c6],
    ['CDU/CSU', 'SPD', 'CDU/CSU', 'SPD', 'CDU/CSU', 'SPD'],
    default=np.nan
)

# 3) Keep only rows where a party was selected
df = df[df['party'].notna()].copy()

# 4) Extract that party’s vote share for each row
df['vote_share'] = df.apply(lambda r: r[r['party']], axis=1)

# 5) Final tidy result
result = df[['Jahr', 'party', 'vote_share']].sort_values('Jahr').reset_index(drop=True)

print(result)


    Jahr    party  vote_share
0   1953  CDU/CSU        45.2
1   1957  CDU/CSU        50.2
2   1961  CDU/CSU        45.3
3   1965  CDU/CSU        47.6
4   1969  CDU/CSU        46.1
5   1972      SPD        45.8
6   1976      SPD        42.6
7   1980      SPD        42.9
8   1983      SPD        38.2
9   1987  CDU/CSU        44.3
10  1990  CDU/CSU        43.8
11  1994  CDU/CSU        41.5
12  1998  CDU/CSU        35.2
13  2002      SPD        38.5
14  2005      SPD        34.2
15  2009  CDU/CSU        33.8
16  2013  CDU/CSU        41.5
17  2017  CDU/CSU        32.9
18  2021  CDU/CSU        24.1
19  2025      SPD        16.4


Next, I am importing GDP growth rate data for the years 1950-2022. Unfortunately, I couldn't find any historic quarterly data on GDP growth. I am sure it exists but either is not publicly available or requires a specific inquiry at a German research institution. Further research can work on data quality. For this project I decided that yearly data will suffice. 

In [520]:
# Importing GDP growth data (G):

import pandas as pd 
G = pd.read_excel("bip-1950-heute.xlsx", header=1)


In [521]:
G.head(100)

Jahr     G
0   1951   9.7
1   1952   9.3
2   1953   8.9
3   1954   7.8
4   1955  12.1
..   ...   ...
67  2018   1.0
68  2019   1.1
69  2020  -3.7
70  2021   2.6
71  2022   1.8

[72 rows x 2 columns]

Here, I manually add GDP growth data for 2023 and 2024: 

In [522]:
# add a line with -0.3% for 2023 and -0.5% for 2024

new_rows = pd.DataFrame([
    {'Jahr': 2023, 'G': -0.3},
    {'Jahr': 2024, 'G': -0.5},
])

G = pd.concat([G, new_rows], ignore_index=True)

G.head(100)


Jahr     G
0   1951   9.7
1   1952   9.3
2   1953   8.9
3   1954   7.8
4   1955  12.1
..   ...   ...
69  2020  -3.7
70  2021   2.6
71  2022   1.8
72  2023  -0.3
73  2024  -0.5

[74 rows x 2 columns]

Next, I am importing the yearly consumer price index for germany or to be more precise it's growth rate, i.e. inflation:

In [523]:
# I am importing Verbraucherpreisindex (Consumer Price Index) data:
import pandas as pd

CPI = pd.read_excel("VPI.xlsx", sheet_name= "Daten", header=4)


In [524]:
CPI.head(100)

Unnamed: 0  Unnamed: 1  Unnamed: 2 Unnamed: 3
0          NaN        1950        -6.4       in %
1          NaN        1951         7.6       in %
2          NaN        1952         2.1       in %
3          NaN        1953        -1.7       in %
4          NaN        1954         0.4       in %
..         ...         ...         ...        ...
70         NaN        2020         0.5       in %
71         NaN        2021         3.1       in %
72         NaN        2022         6.9       in %
73         NaN        2023         5.9       in %
74         NaN        2024         2.2       in %

[75 rows x 4 columns]

In [525]:
# remove first and last column
CPI = CPI.iloc[:, 1:-1]

# Rename remaining columns
CPI.columns = ["Jahr", "Z"]   

CPI.head(100)

Jahr    Z
0   1950 -6.4
1   1951  7.6
2   1952  2.1
3   1953 -1.7
4   1954  0.4
..   ...  ...
70  2020  0.5
71  2021  3.1
72  2022  6.9
73  2023  5.9
74  2024  2.2

[75 rows x 2 columns]

I wanna use Prof. Fair's model as a starting point but since I only have 19 observations I don't want to overfit which is why I am only gonna use inflation and GDP growth data. The incumbent dummy is somewhat integrated into the model due to the fact that we are looking at the vote share of the governing, i.e. the incumbent party even if they don't stay in power:

$$V^{p} = \alpha_{0} + \alpha_{1} G + \alpha_{2} Z + \varepsilon$$

One limitation of the data is that I don't have quarterly data on GDP growth and inflation but only yearly. Next, I am gonna create lead variables of 1-3 years back in time of G and Z to then create a covariance or correlation matrix to get an idea of how those economic indicators might be affecting voting behaviour. 

In [526]:
# Merging dataframes and creating lag variables
df = G.merge(CPI, on="Jahr", how="inner")    
df = df.merge(result, on="Jahr", how="outer")

for i in range(1, 4):
    df[f"G_lag{i}"] = df["G"].shift(i)
    df[f"Z_lag{i}"] = df["Z"].shift(i)

df = df[df["party"].notna()]
df.head(100)

df.head(100)




Jahr         G    Z    party  vote_share   G_lag1  Z_lag1     G_lag2  \
2   1953  8.900000 -1.7  CDU/CSU        45.2  9.30000     2.1   9.700000   
6   1957  6.100000  2.0  CDU/CSU        50.2  7.70000     2.8  12.100000   
10  1961  4.600000  2.5  CDU/CSU        45.3  8.60000     1.6   7.900000   
14  1965  5.400000  3.2  CDU/CSU        47.6  6.70000     2.4   2.800000   
18  1969  7.500000  1.8  CDU/CSU        46.1  5.50000     1.6  -0.300000   
21  1972  4.300000  5.4      SPD        45.8  3.10000     5.2   5.000000   
25  1976  4.900000  4.2      SPD        42.6 -0.90000     6.0   0.900000   
29  1980  1.400000  5.4      SPD        42.9  4.20000     4.1   3.000000   
32  1983  1.600000  3.2      SPD        38.2 -0.40000     5.2   0.500000   
36  1987  1.400000  0.2  CDU/CSU        44.3  2.30000    -0.1   2.300000   
39  1990  5.300000  2.6  CDU/CSU        43.8  3.90000     2.8   3.700000   
43  1994  2.391892  2.7  CDU/CSU        41.5 -0.97685     4.5   1.923077   
47  1998  2.000000  0.8  CDU/CSU        35.2  1.80000     1.9   0.800000   
51  2002 -0.200000  1.4      SPD        38.5  1.70000     2.0   2.900000   
54  2005  0.700000  1.6      SPD        34.2  1.20000     1.6  -0.700000   
58  2009 -5.700000  0.3  CDU/CSU        33.8  1.00000     2.6   3.000000   
62  2013  0.400000  1.5  CDU/CSU        41.5  0.40000     1.9   3.900000   
66  2017  2.700000  1.5  CDU/CSU        32.9  2.20000     0.5   1.500000   
70  2021  2.600000  3.1  CDU/CSU        24.1 -3.70000     0.5   1.100000   
74  2025       NaN  NaN      SPD        16.4 -0.50000     2.2  -0.300000   

    Z_lag2  G_lag3  Z_lag3  
2      7.6     NaN     NaN  
6      1.4     7.8     0.4  
10     0.6     4.5     2.3  
14     3.0     4.7     2.8  
18     1.9     2.8     3.3  
21     3.6     7.5     1.8  
25     6.9     4.8     7.1  
29     2.7     3.3     3.7  
32     6.3     1.4     5.4  
36     2.0     2.8     2.5  
39     1.2     1.4     0.2  
43     5.0     5.1     3.7  
47     1.4     1.5     1.9  
51     1.3     1.9     0.7  
54     1.0    -0.2     1.4  
58     2.3     3.8     1.6  
62     2.2     4.2     1.0  
66     0.5     2.2     1.0  
70     1.4     1.0     1.8  
74     5.9     1.8     6.9

In [527]:
df_corr = df.drop(columns=["party"]).corr()
print(df_corr.to_latex())


\begin{tabular}{lrrrrrrrrrr}
\toprule
 & Jahr & G & Z & vote_share & G_lag1 & Z_lag1 & G_lag2 & Z_lag2 & G_lag3 & Z_lag3 \\
\midrule
Jahr & 1.000000 & -0.711534 & -0.097810 & -0.840480 & -0.789002 & -0.324930 & -0.618384 & -0.227927 & -0.607932 & -0.042927 \\
G & -0.711534 & 1.000000 & 0.061165 & 0.561216 & 0.578390 & 0.067694 & 0.385583 & 0.275050 & 0.295484 & 0.152115 \\
Z & -0.097810 & 0.061165 & 1.000000 & 0.110728 & -0.222688 & 0.617834 & -0.193371 & 0.050449 & 0.349128 & 0.446107 \\
vote_share & -0.840480 & 0.561216 & 0.110728 & 1.000000 & 0.669436 & 0.264332 & 0.555958 & -0.041249 & 0.602743 & -0.267467 \\
G_lag1 & -0.789002 & 0.578390 & -0.222688 & 0.669436 & 1.000000 & -0.131453 & 0.712361 & -0.117098 & 0.427597 & -0.331664 \\
Z_lag1 & -0.324930 & 0.067694 & 0.617834 & 0.264332 & -0.131453 & 1.000000 & 0.015469 & 0.605093 & 0.432323 & 0.494749 \\
G_lag2 & -0.618384 & 0.385583 & -0.193371 & 0.555958 & 0.712361 & 0.015469 & 1.000000 & -0.014678 & 0.713968 & -0.450959 \\
Z_lag2 &

Inflation doesn't seem to have a strong correlation with the vote share of the governing party, not even its lags. In contrast, GDP growth with one lag has a correlation of 0.669, which is not bad. Following these results, I'll include GDP growth lag 1 and not include the inflation rate. I will also add the variable DUR. Then, I do have a model with two predictors. I should not have any more due to the small number of observations and the risk of overfitting. 

In [ ]:

import numpy as np
import statsmodels.api as sm

Y = df["vote_share"]
X = df[["G_lag1", "Z_lag1"]]



X = sm.add_constant(X)

germ_reg = sm.OLS(Y, X).fit(cov_type='HAC',cov_kwds={'maxlags':1})
print(germ_reg.summary())    




\begin{center}
\begin{tabular}{lclc}
\toprule
\textbf{Dep. Variable:}    &   vote\_share    & \textbf{  R-squared:         } &     0.574   \\
\textbf{Model:}            &       OLS        & \textbf{  Adj. R-squared:    } &     0.524   \\
\textbf{Method:}           &  Least Squares   & \textbf{  F-statistic:       } &     15.48   \\
\textbf{Date:}             & Sun, 23 Nov 2025 & \textbf{  Prob (F-statistic):} &  0.000148   \\
\textbf{Time:}             &     17:47:22     & \textbf{  Log-Likelihood:    } &   -61.550   \\
\textbf{No. Observations:} &          20      & \textbf{  AIC:               } &     129.1   \\
\textbf{Df Residuals:}     &          17      & \textbf{  BIC:               } &     132.1   \\
\textbf{Df Model:}         &           2      & \textbf{                     } &             \\
\textbf{Covariance Type:}  &       HAC        & \textbf{                     } &             \\
\bottomrule
\end{tabular}
\begin{tabular}{lcccccc}
                 & \textbf{coef} & \tex

Here I specify a polynomial model where I include Z lag and Z lag squared. 

In [ ]:
import numpy as np
import statsmodels.api as sm

# 1) Build Y
Y = df["vote_share"]

# 2) Create polynomial terms for Z_lag1
df["Z_lag1_sq"] = df["Z_lag1"] ** 2     # square term

# 3) Build X with G_lag1, Z_lag1, Z_lag1^2
X = df[["G_lag1", "Z_lag1", "Z_lag1_sq"]]

# 4) Add constant and fit
X = sm.add_constant(X)
germ_reg = sm.OLS(Y, X).fit(cov_type='HAC',cov_kwds={'maxlags':1})

print(germ_reg.summary())


\begin{center}
\begin{tabular}{lclc}
\toprule
\textbf{Dep. Variable:}    &   vote\_share    & \textbf{  R-squared:         } &     0.639   \\
\textbf{Model:}            &       OLS        & \textbf{  Adj. R-squared:    } &     0.572   \\
\textbf{Method:}           &  Least Squares   & \textbf{  F-statistic:       } &     14.32   \\
\textbf{Date:}             & Sun, 23 Nov 2025 & \textbf{  Prob (F-statistic):} &  8.54e-05   \\
\textbf{Time:}             &     17:46:41     & \textbf{  Log-Likelihood:    } &   -59.901   \\
\textbf{No. Observations:} &          20      & \textbf{  AIC:               } &     127.8   \\
\textbf{Df Residuals:}     &          16      & \textbf{  BIC:               } &     131.8   \\
\textbf{Df Model:}         &           3      & \textbf{                     } &             \\
\textbf{Covariance Type:}  &       HAC        & \textbf{                     } &             \\
\bottomrule
\end{tabular}
\begin{tabular}{lcccccc}
                     & \textbf{coef} & 

Next, I'll create the the DUR variable. 



In [530]:
dur = {
    1953: 1,
    1957: 2,
    1961: 3,
    1965: 4,
    1969: 5,
    1972: 1,
    1976: 2,
    1980: 3,
    1983: 1,
    1987: 2,
    1990: 3,
    1994: 4,
    1998: 5,
    2002: 1,
    2005: 2,
    2009: 1,
    2013: 2,
    2017: 3,
    2021: 4,
    2025: 1
}

df["DUR"] = df["Jahr"].map(dur)
df.head(20)









Jahr         G    Z    party  vote_share   G_lag1  Z_lag1     G_lag2  \
2   1953  8.900000 -1.7  CDU/CSU        45.2  9.30000     2.1   9.700000   
6   1957  6.100000  2.0  CDU/CSU        50.2  7.70000     2.8  12.100000   
10  1961  4.600000  2.5  CDU/CSU        45.3  8.60000     1.6   7.900000   
14  1965  5.400000  3.2  CDU/CSU        47.6  6.70000     2.4   2.800000   
18  1969  7.500000  1.8  CDU/CSU        46.1  5.50000     1.6  -0.300000   
21  1972  4.300000  5.4      SPD        45.8  3.10000     5.2   5.000000   
25  1976  4.900000  4.2      SPD        42.6 -0.90000     6.0   0.900000   
29  1980  1.400000  5.4      SPD        42.9  4.20000     4.1   3.000000   
32  1983  1.600000  3.2      SPD        38.2 -0.40000     5.2   0.500000   
36  1987  1.400000  0.2  CDU/CSU        44.3  2.30000    -0.1   2.300000   
39  1990  5.300000  2.6  CDU/CSU        43.8  3.90000     2.8   3.700000   
43  1994  2.391892  2.7  CDU/CSU        41.5 -0.97685     4.5   1.923077   
47  1998  2.000000  0.8  CDU/CSU        35.2  1.80000     1.9   0.800000   
51  2002 -0.200000  1.4      SPD        38.5  1.70000     2.0   2.900000   
54  2005  0.700000  1.6      SPD        34.2  1.20000     1.6  -0.700000   
58  2009 -5.700000  0.3  CDU/CSU        33.8  1.00000     2.6   3.000000   
62  2013  0.400000  1.5  CDU/CSU        41.5  0.40000     1.9   3.900000   
66  2017  2.700000  1.5  CDU/CSU        32.9  2.20000     0.5   1.500000   
70  2021  2.600000  3.1  CDU/CSU        24.1 -3.70000     0.5   1.100000   
74  2025       NaN  NaN      SPD        16.4 -0.50000     2.2  -0.300000   

    Z_lag2  G_lag3  Z_lag3  Z_lag1_sq  DUR  
2      7.6     NaN     NaN       4.41    1  
6      1.4     7.8     0.4       7.84    2  
10     0.6     4.5     2.3       2.56    3  
14     3.0     4.7     2.8       5.76    4  
18     1.9     2.8     3.3       2.56    5  
21     3.6     7.5     1.8      27.04    1  
25     6.9     4.8     7.1      36.00    2  
29     2.7     3.3     3.7      16.81    3  
32     6.3     1.4     5.4      27.04    1  
36     2.0     2.8     2.5       0.01    2  
39     1.2     1.4     0.2       7.84    3  
43     5.0     5.1     3.7      20.25    4  
47     1.4     1.5     1.9       3.61    5  
51     1.3     1.9     0.7       4.00    1  
54     1.0    -0.2     1.4       2.56    2  
58     2.3     3.8     1.6       6.76    1  
62     2.2     4.2     1.0       3.61    2  
66     0.5     2.2     1.0       0.25    3  
70     1.4     1.0     1.8       0.25    4  
74     5.9     1.8     6.9       4.84    1

Next, I specify the following model: 


$$V^{p} = \alpha_{0} + \alpha_{1} G + \alpha_{2} DUR + \varepsilon$$

In [531]:
df["DUR_dummy"] = (df["DUR"] >= 2).astype(int)


Here I specify a model with G lag and the DUR dummy:

In [532]:

import numpy as np
import statsmodels.api as sm

Y = df["vote_share"]
X = df[["G_lag1", "DUR_dummy"]]

X = sm.add_constant(X)

germ_reg_2 = sm.OLS(Y, X).fit(cov_type='HAC',cov_kwds={'maxlags':1})
print(germ_reg_2.summary())




                            OLS Regression Results                            
Dep. Variable:             vote_share   R-squared:                       0.498
Model:                            OLS   Adj. R-squared:                  0.438
Method:                 Least Squares   F-statistic:                     8.016
Date:                Sun, 23 Nov 2025   Prob (F-statistic):            0.00353
Time:                        17:44:49   Log-Likelihood:                -63.211
No. Observations:                  20   AIC:                             132.4
Df Residuals:                      17   BIC:                             135.4
Df Model:                           2                                         
Covariance Type:                  HAC                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         32.6329      3.853      8.470      0.0

The model shows some explanatory power but also some limitations. With an Adj. R-Squared of just 0.43, the model only explains 43% of the variation in the vote share data. However, a small P-value suggests a strong correlation between G_lag1 and the party's vote share. With a P-value of 0.213, the DUR dummy seems not to have explanatory power for vote share. 







Next, I am testing the predictive power of the model using the Leave-One-Out Cross-Validation (LOOCV) method, which works relatively well for models that are fit on a smaller number of data. I get a relatively high RMSE of 6.9 %, which is a lot and a LOOCV R^2 of 0.262. 

In [533]:
from sklearn.model_selection import LeaveOneOut
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

Y = df["vote_share"]
X = df[["G_lag1", "Z_lag1"]]

loo = LeaveOneOut()
preds = []
truth = []

for train_idx, test_idx in loo.split(X):
    X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
    y_train, y_test = Y.iloc[train_idx], Y.iloc[test_idx]

    model = LinearRegression(fit_intercept=True)
    model.fit(X_train, y_train) 

    preds.append(model.predict(X_test)[0])
    truth.append(y_test.values[0])

# Compute LOOCV error
preds = np.array(preds)
truth = np.array(truth)

rmse = np.sqrt(mean_squared_error(truth, preds))
print("LOOCV RMSE:", rmse)

# Optional R^2 on LOOCV predictions
r2 = 1 - np.sum((truth - preds)**2) / np.sum((truth - truth.mean())**2)
print("LOOCV R^2:", r2)


LOOCV RMSE: 6.066457603793037
LOOCV R^2: 0.4321873983979666


Here I predict the 2021 election based on a model that was trained on data without 2021 and I compare the predicted value to the actual value: 

In [536]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

# 1) Work on a copy to be safe
df_model = df.copy()

# 2) Target variable
Y = df_model["vote_share"].astype(float)

# 3) Create polynomial term for Z_lag1
df_model["Z_lag1_sq"] = df_model["Z_lag1"] ** 2

# 4) Build X with G_lag1, Z_lag1, Z_lag1^2
X = df_model[["G_lag1", "Z_lag1", "Z_lag1_sq"]].astype(float)

# 5) Add constant
X = sm.add_constant(X, has_constant="add")

# 6) Combine and clean (remove NaN and ±inf)
data = pd.concat([Y, X], axis=1)
data = data.replace([np.inf, -np.inf], np.nan).dropna()

# 7) Train/test split: all but last for training, last for prediction
train = data.iloc[:-1]
test  = data.iloc[[-1]]   # keep as DataFrame

Y_train = train["vote_share"]
X_train = train.drop(columns=["vote_share"])

Y_test_actual = test["vote_share"].iloc[0]
X_test        = test.drop(columns=["vote_share"])

# 8) Fit OLS on training data only
germ_reg = sm.OLS(Y_train, X_train).fit(cov_type='HAC',cov_kwds={'maxlags':1})

# If you still want LaTeX summary:
print(germ_reg.summary())

# 9) Predict the last observation
Y_pred_last = germ_reg.predict(X_test)

print("\nActual last vote_share:", Y_test_actual)
print("Predicted last vote_share:", float(Y_pred_last))


                            OLS Regression Results                            
Dep. Variable:             vote_share   R-squared:                       0.712
Model:                            OLS   Adj. R-squared:                  0.654
Method:                 Least Squares   F-statistic:                     17.18
Date:                Sun, 23 Nov 2025   Prob (F-statistic):           4.07e-05
Time:                        17:45:39   Log-Likelihood:                -49.854
No. Observations:                  19   AIC:                             107.7
Df Residuals:                      15   BIC:                             111.5
Df Model:                           3                                         
Covariance Type:                  HAC                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         34.2840      3.559      9.633      0.0

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=19
  res = hypotest_fun_out(*samples, **kwds)
/var/folders/x5/gq2pkkwj427ctp4rg1stnj180000gn/T/ipykernel_70400/1247801857.py:44: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  print("Predicted last vote_share:", float(Y_pred_last))
